common/bb.1
.TH BB 1 "Version 4.2.3:  4 Feb 2009" "Xymon"
.SH NAME
bb \- Xymon client communication program
.SH SYNOPSIS
.B "bb [options] RECIPIENT message"

.SH DESCRIPTION
.I bb(1)
is the client program used to communicate with a
Xymon server. It is frequently used by Xymon
client systems to send in status messages and pager
alerts on local tests.

In Xymon, the bb program is also used for administrative
purposes, e.g. to rename or delete hosts, or to disable
hosts that are down for longer periods of time.

.SH OPTIONS AND PARAMETERS
.IP "--debug"
Enable debugging. This prints out details about how the
connection to the BBDISPLAY server is being established.

.IP "--proxy=http://PROXYSERVER:PROXYPORT/"
When sending the status messages via HTTP, use this server
as an HTTP proxy instead of connecting directly to the BBDISPLAY
server.

.IP "--timeout=N"
Specifies the timeout for connecting to the Xymon server, in
seconds. The default is 5 seconds.

.IP "--response"
The bb utility normally knows when to expect a response 
from the server, so this option is not required. However,
it will cause any response from the server to be displayed.

.IP "--merge"
Merge the command line message text with the data provided
on standard input, and send the result to the Xymon server.
The message text provided on the command line becomes the
first line of the merged message.

.IP "RECIPIENT"
The \fBRECIPIENT\fR parameter defines which server receives
the message. If RECIPIENT is given as "0.0.0.0", then the
message is sent to all of the servers listed in the BBDISPLAYS
or BBPAGERS environment variable (for "status" and "page" messages,
respectively). 

Usually, a client will use "$BBDISP" for the \fBRECIPIENT\fR 
parameter, as this is defined for the client scripts
to automatically contain the correct value.

The \fBRECIPIENT\fR parameter may be a URL for a webserver
that has the bbmessage.cgi or similar script installed. This
tunnels the Xymon messages to the BBDISPLAY server 
using standard HTTP protocol. The 
.I bbmessage.cgi(8)
CGI tool (included in Xymon) must be installed on the webserver 
for the HTTP transport to work.
.br

.IP MESSAGE
The \fBmessage\fR parameter is the message to be sent across
to the Xymon server. Messages must be enclosed in quotes,
but by doing so they can span multiple lines. The maximum size
of a message is defined by the maximum allowed length of your
shell\(aqs command-line, and is typically 8-32 KB. 

If you need to send longer status messages, you can specify "@" 
as the message: bb will then read the status message from its
stdin.

.SH XYMON MESSAGE SYNTAX

This section lists the most commonly used messages in the Xymon
protocol.

Each message must begin with one of the Xymon commands. Where
a HOSTNAME is specified, it must have any dots in the hostname changed
to commas if the Xymon FQDN setting is enabled (which is the default).
So the host "www.foo.com", for example, would report as "www,foo,com".

.IP "status[+LIFETIME][/group:GROUP] HOSTNAME.TESTNAME COLOR <additional text>"
This sends in a status message for a single test (column) on a single host.
TESTNAME is the name of the column where this test will show up; any
name is valid except that using dots in the testname will not work.
COLOR must be one of the valid colors: "green", "yellow", "red" or "clear".
The colors "blue" and "purple" - although valid colors - should not be sent in a
status message, as these are handled specially by the Xymon server.
As a special case (for supporting older clients), "client" can be
used as the name of the color. This causes the status message to be
handled by Xymon as a "client" data message, and the TESTNAME
parameter is used as the "collector id".
.br
The "additional text" normally includes a local timestamp and a summary
of the test result on the first line. Any lines following the first one
are free-form, and can include any information that may be useful to
diagnose the problem being reported.
.br
The LIFETIME defines how long this status is valid after being received
by the Xymon server. The default is 30 minutes, but you can set any
period you like. E.g. for a custom test that runs once an hour, you will
want to set this to at least 60 minutes - otherwise the status will go
purple after 30 minutes. It is a good idea to set the LIFETIME to
slightly longer than the interval between your tests, to allow for variations
in the time it takes your test to complete. The LIFETIME is in minutes,
unless you add an "h" (hours), "d" (days) or "w" (weeks) immediately after
the number, e.g. "status+5h" for a status that is valid for 5 hours.
.br
The GROUP option is used to direct alerts from the status to a specific group.
It is currently used for status generated from the Xymon clients\(aq data,
e.g. to direct alerts for a "procs" status to different people, depending on
exactly which process is down.

.IP "notify HOSTNAME.TESTNAME <message text>"
This triggers an informational message to be sent to those who 
receive alerts for this HOSTNAME+TESTNAME combination, 
according to the rules defined in
.I hobbit-alerts.cfg(5)
This is used by the 
.I hobbit-enadis.cgi(1)
tool to notify people about hosts being disabled or 
enabled, but can also serve as a general way of notifying 
server administrators.

.IP "data HOSTNAME.DATANAME<newline><additional text>"
The "data" message allows tools to send data about a host, without
it appearing as a column on the Xymon webpages. This
is used, for example, to report statistics about a host, e.g. vmstat data, which
does not in itself represent something that has a red, yellow or
green identity. It is used by RRD bottom-feeder modules, among
others. In Xymon, data messages are by default processed only by the
.I hobbitd_rrd(8)
module. If you want to handle data-messages using an external application,
you may want to enable the 
.I hobbitd_filestore(8)
module for data-messages, to store data-messages in a format compatible
with how the Big Brother daemon does.

.IP "disable HOSTNAME.TESTNAME DURATION <additional text>"
Disables a specific test for DURATION minutes. This will cause the
status of this test to be listed as "blue" on the BBDISPLAY server,
and no alerts for this host/test will be generated. If DURATION is
given as a number followed by s/m/h/d, it is interpreted as being
in seconds/minutes/hours/days respectively.
.BR
To disable a test until it becomes OK, use "-1" as the DURATION.
.BR
To disable all tests for a host, use an asterisk "*" for TESTNAME.

.IP "enable HOSTNAME.TESTNAME"
Re-enables a test that had been disabled.

.IP "query HOSTNAME.TESTNAME"
Query the BBDISPLAY server for the latest status reported for this
particular test. If the host/test status is known, the response is
the first line of the status report - the current color will be the
first word on the line. Additional lines of text that might be 
present on the status message cannot be retrieved.
.br
This allows any Xymon client to determine the status of a particular
test, whether it is one pertaining to the host where the client
is running, some other host, or perhaps the result of a combined
test from multiple hosts managed by
.I bbcombotest(1)
This will typically be useful to Xymon client extension scripts, that
need to determine the status of other hosts, for example, to decide if an
automatic recovery action should be initiated.

.IP "config FILENAME"
Retrieve one of the Xymon configuration files from the
server. This command allows a client to pull files from the
$BBHOME/etc/ directory on the server, allowing for semi-automatic
updates of the client configuration. Since the configuration files 
are designed to have a common file for the configuration of all hosts 
in the system - and this is in fact the recommended way of configuring 
your clients - this makes it easier to keep the configuration 
files synchronized.

.IP "drop HOSTNAME"
Removes all data stored about the host HOSTNAME. It is assumed that you
have already deleted the host from the bb-hosts configuration file.

.IP "drop HOSTNAME TESTNAME"
Remove data about a single test (column).

.IP "rename OLDHOSTNAME NEWHOSTNAME"
Rename all data for a host that has had its name changed. You should do this
before changing the hostname in the bb-hosts configuration file.

.IP "rename HOSTNAME OLDTESTNAME NEWTESTNAME"
Rename data about a single test (column).

.IP "hobbitdlog HOSTNAME.TESTNAME"
Retrieve the Xymon status-log for a single test. The first line of the
response contains a series of fields separated by a pipe-sign:
.sp
.BR hostname
The name of the host
.sp
.BR testname
The name of the test
.sp
.BR color
Status color (green, yellow, red, blue, clear, purple)
.sp
.BR testflags
For network tests, the flags indicating details about the test (used by bbgen).
.sp
.BR lastchange
Unix timestamp when the status color last changed.
.sp
.BR logtime
Unix timestamp when the log message was received.
.sp
.BR validtime
Unix timestamp when the log message is no longer valid (it goes purple at this time).
.sp
.BR acktime
Either -1 or Unix timestamp when an active acknowledgement expires.
.sp
.BR disabletime
Either -1 or Unix timestamp when the status is no longer disabled.
.sp
.BR sender
IP address where the status was received from.
.sp
.BR cookie
Either -1 or the cookie value used to acknowledge an alert.
.sp
.BR ackmsg
Empty or the acknowledgment message sent when the status was acknowledged.
Newline, pipe-signs and backslashes are escaped with a backslash, C-style.
.sp
.BR dismsg
Empty or the message sent when the status was disabled.
Newline, pipe-signs and backslashes are escaped with a backslash, C-style.
.sp
After the first line comes the full status log in plain text format.

.IP "hobbitdxlog HOSTNAME.TESTNAME"
Retrieves an XML string containing the status log as with the 
"hobbitdlog" command.

.IP "hobbitdboard [CRITERIA] [fields=FIELDLIST]"
Retrieves a summary of the status of all known tests available to
the Xymon daemon. 

By default - if no CRITERIA is provided - it returns one line for all 
status messages that are found in Xymon. You can filter the response
by selecting a page, a host, a test or a color - wildcards are not 
supported, so you can pick only one page, host, test or color. 
.sp
.BR page=PAGEPATH
Include only tests from hosts found on the PAGEPATH page in the bb-hosts
file.
.sp
.BR host=HOSTNAME
Include only tests from the host HOSTNAME
.sp
.BR test=TESTNAME
Include only tests with the testname TESTNAME
.sp
.BR color=COLORNAME
Include only tests where the status color is COLORNAME
.sp
You can filter on, for example, both a hostname and a testname. 

The response is one line for each status that matches the CRITERIA,
or all statuses if no criteria is specified. The line is composed of
a number of fields, separated by a pipe-sign. You can select which
fields to retrieve by listing them in the FIELDLIST. The following
fields are available:
.sp
.BR hostname
The name of the host
.sp
.BR testname
The name of the test
.sp
.BR color
Status color (green, yellow, red, blue, clear, purple)
.sp
.BR flags
For network tests, the flags indicating details about the test (used by bbgen).
.sp
.BR lastchange
Unix timestamp when the status color last changed.
.sp
.BR logtime
Unix timestamp when the log message was received.
.sp
.BR validtime
Unix timestamp when the log message is no longer valid (it goes purple at this time).
.sp
.BR acktime
Either -1 or Unix timestamp when an active acknowledgement expires.
.sp
.BR disabletime
Either -1 or Unix timestamp when the status is no longer disabled.
.sp
.BR sender
IP address where the status was received from.
.sp
.BR cookie
Either -1 or the cookie value used to acknowledge an alert.
.sp
.BR line1
First line of status log.
.sp
.BR ackmsg
Empty (if no acknowledgement is active), or the text of the acknowledge
message.
.sp
.BR dismsg
Empty (if the status is currently enabled), or the text of the disable message.
.sp
.BR msg
The full text of the current status message.

The ackmsg, dismsg and msg fields have certain characters encoded: Newline
is "\\n", TAB is "\\t", carriage return is "\\r", a pipe-sign is "\\p", 
and a backslash is "\\\\".

If the "fields" parameter is omitted, a default set of
hostname,testname,color,flags,lastchange,logtime,validtime,acktime,disabletime,sender,cookie,line1
is used.

.IP "hobbitdxboard"
Retrieves an XML string with the summary of all status logs
as for the "hobbitdboard" command.

.IP "download FILENAME"
Download a file from the Xymon server\(aqs download directory.

.IP "client[/COLLECTORID] HOSTNAME.OSTYPE [HOSTCLASS]"
Used to send a "client" message to the Xymon server. Client messages
are generated by the Xymon client; when sent to the Xymon server they
are matched against the rules in the
.I hobbit-clients.cfg(5)
configuration file, and status messages are generated for the client-side
tests.
The COLLECTORID is used when sending client-data that are additions
to the standard client data. The data will be concatenated with the
normal client data.

.IP "clientlog HOSTNAME [section=SECTIONNAME[,SECTIONNAME...]]"
Retrieves the current raw client message last sent by HOSTNAME. The optional
"section" filter is used to select specific sections of the client data.

.IP "ping"
Attempts to contact the Xymon server. If successful, the Xymon server version ID
is reported.

.IP "pullclient"
This message is used when fetching client data via the "pull" mechanism implemented by
.I hobbitfetch(8)
and
.I msgcache(8)
for clients that cannot connect directly to the Xymon server.

.IP "ghostlist"
Report a list of \fBghost\fR clients seen by the Xymon server. Ghosts are systems
that report data to the Xymon server, but are not listed in the bb-hosts file.

.IP "schedule [TIMESTAMP COMMAND]"
Schedules a command sent to the Xymon server for execution at a later time. E.g.
used to schedule disabling of a host or service at sometime in the future. COMMAND
is a complete Xymon command such as the ones listed above. TIMESTAMP is the
Unix epoch time when the command will be executed.
.br
If no parameters are given, the currently scheduled tasks are listed in the response.
The response is one line per scheduled command, with the job-id, the time when the
command will be executed, the IP address from which this was sent, and the full command
string.
.br
To cancel a previously scheduled command, \fB"schedule cancel JOBID"\fR can be
used. JOBID is a number provided as the first item in the output from the schedule list.

.IP "notes FILENAME"
The message text will be stored in $BBHOME/notes/FILENAME which is then used as
hyperlinks from hostnames or column names. This requires that the "bbnotes" 
task is enabled in hobbitlaunch.cfg (it is disabled by default). FILENAME 
cannot contain any directory path - these are stripped automatically.

.IP "usermsg ID"
These messages will be relayed directly to modules listening on the "user"
channel of the Xymon daemon. This is intended for custom communication
between client-side modules and the Xymon server.

.IP "modify HOSTNAME.TESTNAME COLOR SOURCE CAUSE"
Modify the color of a specific status, without generating a complete
status message. This is for backend processors (e.g. RRD graphs)
that can override the color of a status based on some criteria
determined outside the normal flow of a status. E.g. the normal
"conn" status may appear to be green since it merely checks on
whether a host can be ping'ed or not; the RRD handler can then
use a "modify" command to override this is the actual ping
responsetime exceeds a given threshold. (See the "DS" configuration
setting in 
.I hobbit-clients.cfg(5)
for how to do this). SOURCE is some identification of the module
that generates the "modify" message - future modifications must
use the same source. There may be several sources that modify
the same status (the most severe status then becomes the actual
color of the status). CAUSE is a one-line text string explaining
the reason for overriding the normal status color - it will be
displayed on the status webpage.


.SH EXAMPLE

Send a normal status message to the BBDISPLAY server, using the
standard Xymon protocol on TCP port 1984:
.br
   $ $BB $BBDISP "status www,foo,com.http green \(gadate\(ga Web OK"

Send the same status message, but using HTTP protocol via the
webserver\(aqs bbmessage.cgi script:
.br
   $ $BB http://bb.foo.com/cgi-bin/bbmessage.cgi "status www,foo,com.http green \(gadate\(ga Web OK"

Use "query" message to determine the color of the "www" test, and
restart Apache if it is red:
.br

   $ WWW=\(ga$BB $BBDISP "query www,foo,com.www" | awk \(aq{print $1}\(aq\(ga
   $ if [ "$WWW" = "red" ]; then /etc/init.d/apache restart; fi

Use "config" message to update the local bb-dftab file (but only
if we get a response):
.br

   $ $BB $BBDISP "config bb-dftab" >/tmp/bb-dftab.new
   $ if [ -s /tmp/bb-dftab.new ]; then 
       mv /tmp/bb-dftab.new $BBHOME/etc/bb-dftab
     fi

Send a very large status message that has been built in the
file "statusmsg.txt". Instead of providing it on the command-line, 
pass it via stdin to the bb command:

   $ cat statusmsg.txt | $BB $BBDISP "@"

.SH NOTES
This man-page describes the bb client program provided
as part of Xymon. This implementation provides features 
not present in the standard Big Brother bb client - 
specifically, the support for sending messages over HTTP, 
and many commands such as "query" and "config" are not part 
of the bb client shipped with Big Brother.

.SH "SEE ALSO"
bbcombotest(1), bb-hosts(5), hobbitserver.cfg(5), xymon(7)



Changes
Changes from 4.3.0 beta 1 -> 4.3.0 beta 2 (24 Apr 2009)
=======================================================
* New "--shuffle" option for bbtest-net to run network tests
  in a random order.
* Client startup script now exports important environment variables,
  so they are actually used in systems with a traditional shell.
* hobbitlaunch no longer crashes if there are no tasks
* Client configure script includes librt for clock_gettime()
* New client support for mainframes: z/OS, z/VM, z/VSE
* Enhanced eventlog and top-changing hosts webpages.
* Revert debian package pathnames back to use "hobbit", so
  updates from 4.2.x will actually work.
* Ghostlist options in hobbitcgi.cfg had no effect because
  of typo in setting name.
* "data" messages could crash hobbitd.
* Debug output from hobbitd_channel now logs only the relevant
  data instead of the full message.
* trimhistory now informs the history module to re-open the
  "allevents" file after trimming it.
* devmon template fix
* New "rrdcachectl" utility included.
* Fixed sorting routine (affected holiday list and others)
* Fix generic crash in communications module between Xymon
  programs, where an empty response message would crash caller.

Changes from 4.2.3 -> 4.3.0 beta 1 (09 Feb 2009)
================================================
Core changes:
* New API's for loadhosts and sendmessage, in preparation for
  the full 5.0 changes.
* Always use getcurrenttime() instead of time().
* Support for defining holidays as non-working days in alerts and
  SLA calculations.
* Hosts which appear on multiple pages in the web display can
  use any page they are on in the alerting rules and elsewhere.
* Worker modules (RRD, client-data parsers etc) can operate on
  remote hosts from the hobbitd daemon, for load-sharing.
* Various bugfixes collected over time.

Network test changes:
* Merged new network tests from trunk: SOAP-over-HTTP,
  SSL minimum cipher strength
* Changed network test code to always report a validity period
  for network tests, so it it possible to run network tests less
  often than every 30 minutes (e.g. once an hour).
* Make the content-type setting in HTTP POST tests configurable.
* Make the source-address used for TCP tests configurable.
* Make the acceptable HTTP result codes configurable.
* Use and save HTTP session cookies.

Web changes
* Support generic drop-down lists in templates.
* "NOCOLUMNS" changed to work for all columns.
* New "group-sorted" definition to auto-sort hosts in a group
* Use browser tooltips for host comments
* "Compact" status allows several statuses to appear as a single
  status on the overview webpages.
* Trends page can select the time period to show. Buttons provided
  for the common selections.
* Ghost list report now lists possible candidates for a ghost, 
  based on IP-address or unqualified hostname.

Report changes
* Number of outages as SLA parameter

Miscellaneous
* hobbitlaunch support for running tasks only on certain hosts,
  and for a maximum time.
* Alert script get a unique ID for each alert.



Changes from 4.2.2 -> 4.2.3 (09 Feb 2008)
=========================================
* Time-out code changed to use clock_gettime() with CLOCK_MONOTONIC
* Bugfix for hobbitd/hobbitd_worker communication going out-of-sync
  resulting in "garbled data" being logged and worker modules 
  stopping.
* NCV module now works with negative numbers.
* Several bugfixes in DNS lookup code - could lead to crashes when
  performing DNS tests.
* Switch to C-ARES 1.6.0 - drop support for older versions.
* Run more TCP tests in parallel by not waiting for very slow
  connections to complete before starting new ones.
* Added "hostlist" web utility for spreadsheet-reporting of the
  hosts in Hobbit.


Changes from 4.2.0 -> 4.2.2 (01 Dec 2008)
=========================================

Changelog extracted from Subversion.


------------------------------------------------------------------------
r5935 | storner | 2008-11-26 12:57:11 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   A /branches/4.2.0 (from /trunk:5040)

Create branch for maintenance of the 4.2.x release.


------------------------------------------------------------------------
r5936 | storner | 2008-11-26 13:13:50 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/common/bb-hosts.5

The ability to set a DOWNTIME for an individual test was mentioned in the Changes file, but not documented in the bb-hosts man-page.


------------------------------------------------------------------------
r5937 | storner | 2008-11-26 13:14:48 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/hobbitd/hobbitd_client.c

The hobbitd_client program would crash when running in test-mode, i.e. launched with the --test option to test disk-, procs- and ports-settings. This only applies to running it in test-mode, normal operation is not affected.


------------------------------------------------------------------------
r5938 | storner | 2008-11-26 13:15:40 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/lib/headfoot.c
   M /branches/4.2.0/web/bb-datepage.c
   M /branches/4.2.0/web/bb-eventlog.c
   M /branches/4.2.0/web/bb-findhost.c
   M /branches/4.2.0/web/bb-rep.c
   M /branches/4.2.0/web/bb-snapshot.c
   M /branches/4.2.0/web/hobbit-enadis.c
   M /branches/4.2.0/web/hobbit-nkedit.c

Some of the web page would generate an extra line with "Content-type: text/html" at the top of the webpage, although this would most often be hidden by the menubar.


------------------------------------------------------------------------
r5939 | storner | 2008-11-26 13:16:21 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/client/runclient.sh

Using extra options with the client-side "runclient.sh" script when performing a restart command fails to pass the extra options to the final command. 


------------------------------------------------------------------------

common/bb.c
/*----------------------------------------------------------------------------*/
/* Hobbit communications tool.                                                */
/*                                                                            */
/* This is used to send a single message using the Hobbit/BB protocol to the  */
/* Hobbit server.                                                             */
/*                                                                            */
/* Copyright (C) 2002-2009 Henrik Storner <henrik@hswn.dk>                    */
/*                                                                            */
/* This program is released under the GNU General Public License (GPL),       */
/* version 2. See the file "COPYING" for details.                             */
/*                                                                            */
/*----------------------------------------------------------------------------*/

static char rcsid[] = "$Id$";

#include <sys/types.h>
#include <sys/stat.h>
#include <ctype.h>
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <errno.h>
#include <limits.h>

#include "libbbgen.h"

int main(int argc, char *argv[])
{
	int timeout = BBTALK_TIMEOUT;
	int result = 1;
	int argi;
	int showhelp = 0;
	char *recipient = NULL;
	strbuffer_t *msg = newstrbuffer(0);
	FILE *respfd = stdout;
	char *envarea = NULL;
	sendreturn_t *sres;
	int wantresponse = 0, mergeinput = 0;

	for (argi=1; (argi < argc); argi++) {
		if (strcmp(argv[argi], "--debug") == 0) {
			debug = 1;
		}
		else if (strncmp(argv[argi], "--proxy=", 8) == 0) {
			char *p = strchr(argv[argi], '=');
			setproxy(p+1);
		}
		else if (strcmp(argv[argi], "--help") == 0) {
			showhelp = 1;
		}
		else if (strcmp(argv[argi], "--version") == 0) {
			fprintf(stdout, "Hobbit version %s\n", VERSION);
			return 0;
		}
		else if (strcmp(argv[argi], "--str") == 0) {
			respfd = NULL;
		}
		else if (strncmp(argv[argi], "--out=", 6) == 0) {
			char *fn = argv[argi]+6;
			respfd = fopen(fn, "wb");
		}
		else if (strncmp(argv[argi], "--timeout=", 10) == 0) {
			char *p = strchr(argv[argi], '=');
			timeout = atoi(p+1);
		}
		else if (argnmatch(argv[argi], "--env=")) {
			char *p = strchr(argv[argi], '=');
			loadenv(p+1, envarea);
		}
		else if (argnmatch(argv[argi], "--area=")) {
			char *p = strchr(argv[argi], '=');
			envarea = strdup(p+1);
		}
		else if (strcmp(argv[argi], "--merge") == 0) {
			mergeinput = 1;
		}
		else if (strcmp(argv[argi], "--response") == 0) {
			wantresponse = 1;
		}
		else if (strcmp(argv[argi], "-?") == 0) {
			showhelp = 1;
		}
		else if ((*(argv[argi]) == '-') && (strlen(argv[argi]) > 1)) {
			fprintf(stderr, "Unknown option %s\n", argv[argi]);
		}
		else {
			/* No more options - pickup recipient and msg */
			if (recipient == NULL) {
				recipient = argv[argi];
			}
			else if (STRBUFLEN(msg) == 0) {
				msg = dupstrbuffer(argv[argi]);
			}
			else {
				showhelp=1;
			}
		}
	}

	if ((recipient == NULL) || (STRBUFLEN(msg) == 0) || showhelp) {
		fprintf(stderr, "Hobbit version %s\n", VERSION);
		fprintf(stderr, "Usage: %s [--debug] [--merge] [--proxy=http://ip.of.the.proxy:port/] RECIPIENT DATA\n", argv[0]);
		fprintf(stderr, "  RECIPIENT: IP-address, hostname or URL\n");
		fprintf(stderr, "  DATA: Message to send, or \"-\" to read from stdin\n");
		return 1;
	}

	if (strcmp(STRBUF(msg), "-") == 0) {
		strbuffer_t *inpline = newstrbuffer(0);
		sres = newsendreturnbuf(0, NULL);

		initfgets(stdin);
		while (unlimfgets(inpline, stdin)) {
			result = sendmessage(STRBUF(inpline), recipient, timeout, sres);
			clearstrbuffer(inpline);
		}

		return result;
	}

	if (mergeinput || (strcmp(STRBUF(msg), "@") == 0)) {
		strbuffer_t *inpline = newstrbuffer(0);

		if (mergeinput) 
			/* Must add a new-line before the rest of the message */
			addtostrbuffer(msg, "\n");
		else
			/* Clear input buffer, we'll read it all from stdin */
			clearstrbuffer(msg);

		initfgets(stdin);
		while (unlimfgets(inpline, stdin)) addtostrbuffer(msg, inpline);
		freestrbuffer(inpline);
	}

	if      (strncmp(STRBUF(msg), "query ", 6) == 0) wantresponse = 1;
	else if (strncmp(STRBUF(msg), "client ", 7) == 0) wantresponse = 1;
	else if (strncmp(STRBUF(msg), "config ", 7) == 0) wantresponse = 1;
	else if (strncmp(STRBUF(msg), "download ", 9) == 0) wantresponse = 1;
	else if (strncmp(STRBUF(msg), "hobbitdlog ", 11) == 0) wantresponse = 1;
	else if (strncmp(STRBUF(msg), "hobbitdxlog ", 12) == 0) wantresponse = 1;
	else if (strncmp(STRBUF(msg), "hobbitdboard", 12) == 0) wantresponse = 1;
	else if (strncmp(STRBUF(msg), "hobbitdxboard", 13) == 0) wantresponse = 1;
	else if (strncmp(STRBUF(msg), "schedule", 8) == 0) wantresponse = 1;
	else if (strncmp(STRBUF(msg), "clientlog ", 10) == 0) wantresponse = 1;
	else if (strncmp(STRBUF(msg), "hostinfo", 8) == 0) wantresponse = 1;
	else if (strncmp(STRBUF(msg), "ping", 4) == 0) wantresponse = 1;
	else if (strncmp(STRBUF(msg), "pullclient", 10) == 0) wantresponse = 1;
	else if (strncmp(STRBUF(msg), "ghostlist", 9) == 0) wantresponse = 1;
	else if (strncmp(STRBUF(msg), "multisrclist", 12) == 0) wantresponse = 1;

	sres = newsendreturnbuf(wantresponse, respfd);
	result = sendmessage(STRBUF(msg), recipient, timeout, sres);

	if (sres->respstr) printf("Buffered response is '%s'\n", STRBUF(sres->respstr));

	return result;
}

